### ISSUE TITLE
Seq2Seq: Option to not store whole dataset in memory

### ISSUE THREAD

**Issue by sshleifer** [huggingface/transformers#5476]
https://github.com/huggingface/transformers/blob/ef0e9d806c51059b07b98cb0279a20d3ba3cbc1d/examples/seq2seq/utils.py#L93

**Comment by Pradhy729** [comment]
Do you still need help? I can help out and contribute here.

**Comment by sshleifer** [comment]
Yes that would be super helpful! The goal is to avoid using so much CPU memory here:
https://github.com/huggingface/transformers/blob/353b8f1e7a7361c0afd9e391381bc226b4a5ca8f/examples/seq2seq/utils.py#L101

by only reading a few batches from disk at a time instead of all at once. Will probably require saving in a different format.
Maybe something like this:
https://github.com/pytorch/fairseq/blob/f0a61a2774aff2efbc1adb0b5daee346a8401605/fairseq/data/data_utils.py#L55

Let me know if you need more info!

**Comment by Pradhy729** [comment]
Great! 
My idea is to lazily read & encode just the required line numbers from the file when `__getitem__` is called with an index. For this we could create a map of {example number: line numbers} to read. Let me know what you think.

**Comment by sshleifer** [comment]
Sounds reasonable. What does fairseq do?

**Comment by sshleifer** [comment]
Your approach sounds good, feel free to get started.

**Comment by Pradhy729** [comment]
Great, I will get started

**Comment by Pradhy729** [comment]
Couple of questions : I plan to get rid of the self.source variable; 
1. can I get rid of this property?https://github.com/huggingface/transformers/blob/353b8f1e7a7361c0afd9e391381bc226b4a5ca8f/examples/seq2seq/utils.py#L146-L148

2. Any ideas on how to use the sampler without the full dataset? In general shuffling and sampling may be limited with lazy datasets: although you should be able to use a random sampler in your loader.
https://github.com/huggingface/transformers/blob/353b8f1e7a7361c0afd9e391381bc226b4a5ca8f/examples/seq2seq/utils.py#L155

**Comment by Pradhy729** [comment]
@sshleifer I have a working solution that works for the rest of the training loop and passes the tests. See [here](https://github.com/huggingface/transformers/compare/master...Pradhy729:lazy_loading_seq2seq)
Just need input on my points above. Let me know.

**Comment by sshleifer** [comment]
1) You can get rid of `src_lens` and `tgt_lens`, they are unused afaict

2) I would suggest trying to store the len of each example, (tokenized or untokenized), passing that to `SortishSampler` instead of `self.source`, and then changing
`def key(self, i): len(self.data[i])` -> `def key(self, i): self.data[i]`.

https://github.com/huggingface/transformers/blob/7e86d070c0bbed949b5c922f914f0fec44af72d4/examples/seq2seq/utils.py#L203.

**Comment by sshleifer** [comment]
Like https://github.com/huggingface/transformers/pull/5818

**Comment by Pradhy729** [comment]
Got it. Good idea!
